oca/api.py

```diff
@@ -53,45 +53,31 @@ def method(self):
             instantiated with an iterable of the decorated method results.
 
     Raises:
-        ValueError: If no container type is provided, and a value is returned
-            from the decorated method.
-        TypeError: If an incorrect data type is provided as the `container`.
-            Valid arguments are `None`, instances of `type`, or a callable.
+        AssertionError: If no container type is provided, and a value is
+            returned from the decorated method. This will only be raised if
+            ``PYTHONOPTIMIZE`` is less than 2.
+        AssertionError: If an incorrect data type is provided as `container`.
+            Valid arguments are `None` or instances of `type`. This will only
+            be raised if ``PYTHONOPTIMIZE`` is less than 2.
 
     Returns:
-        list: Of the return values from the decorated method, in the same
-            order as the original ``self`` RecordSet.
+        mixed: A new `container`, instantiated with the results of the
+            decorated method - iterated so that `self` is a singleton.
     """
 
-    method = None
-
-    if container is not None and not isinstance(container, type):
-        if callable(container):
-            method = container
-            container = None
-        else:
-            raise TypeError(
-                'An incorrect argument was provided to the `foreach` api.',
-            )
+    assert container is None or isinstance(container, type)
```

Comment:
Hmm I don't like this. Not always it must be a `type`, but simply a `callable`, such as `iter`. I think you can delete this assertion, since those `if`s are already below.